Fix for "No warning or notification displayed while changing the only admin user privilege from admin to normal. #5522"

soniya20 opened the pull request:
##### SUMMARY
While changing permissions of admin user from Administrator to any other type a notification or warning message should have displayed on screen. But in code there is no functionality exist for the same. 
To add pop–up message in code we have to write a function in “awx/awx/ui/client/src/users/edit/users-edit.controller.js” file. This file is responsible to edit the user’s functionalities. We have added the below conditional check in this file:
if (user_obj.is_superuser === true && $scope.is_superuser === false )

Addresses Issues:
No warning or notification displayed while changing the only admin user privilege from admin to normal. #5522


##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - UI


##### AWX VERSION
AWX- 9.3.0


##### ADDITIONAL INFORMATION
We have freeze the default user’s permission which have user id ‘1’, so that no user can change the default user’s permissions. And also added a pop-up message while changing the superuser permissions.

soniya20 committed:
No warning or notification displayed while changing the only admin user privilege from admin to normal. #5522

We can successfully change the permissions of default admin user from Administrator to any other type which will cause into that there will be no Administrator user present to access the AWX.  This could lead to account hijacking considering the fact that any user with administrator role can modify the default admin user.
That is why we have freeze the default user’s permission which have user id ‘1’, so that no user can change the default user’s permissions.

soniya20 committed:
We have freeze the default user’s permission which have user id ‘1’, so that no user can change the default user’s permissions.

While changing permissions of admin user from Administrator to any other type a notification or warning message should have displayed on screen. But in code there is no functionality exist for the same. 
To add pop–up message in code we have to write a function in “awx/awx/ui/client/src/users/edit/users-edit.controller.js” file. This file is responsible to edit the user’s functionalities. We have added the below conditional check in this file:
if (user_obj.is_superuser === true && $scope.is_superuser === false )

ryanpetrello commented:
@soniya20 thanks for contributing this - I'll see what @mabashian and @jakemcdermott thing about it.  One thing to note is that the AWX UI is currently in the midst of a massive rewrite, and _this_ is the _old_ implementation (which is going away in the next few months).

jakemcdermott commented on awx/ui/client/src/users/edit/users-edit.controller.js:
```diff
@@ -159,19 +159,48 @@ export default ['$scope', '$rootScope', '$stateParams', 'UserForm', 'Rest',
+                alert(user_obj.is_superuser);
+                console.log(user_obj.is_superuser);
+                alert($scope.is_superuser);
+                console.log($scope.is_superuser);
```
Please remove these alerts and logs.

jakemcdermott commented on awx/ui/client/src/users/edit/users-edit.controller.js:
```diff
@@ -159,19 +159,48 @@ export default ['$scope', '$rootScope', '$stateParams', 'UserForm', 'Rest',
+                alert($scope.is_superuser);
+                console.log($scope.is_superuser);
+                if (user_obj.is_superuser === true && $scope.is_superuser === false ) {
+                var action = function() {
```
Please indent the block of code within this `if` statement.

jakemcdermott commented on awx/ui/client/src/users/users.form.js:
```diff
@@ -99,7 +99,7 @@ export default ['i18n', function(i18n) {
                     ngModel: 'user_type',
                     ngShow: 'current_user["is_superuser"]',
-                    ngDisabled: '!(user_obj.summary_fields.user_capabilities.edit || canAdd)'
+                    ngDisabled: 'user_obj.id == 1 || !(user_obj.summary_fields.user_capabilities.edit || canAdd)'
```
We can't assume a default user. It is not necessarily the case that the user with an `id` of 1 is the last remaining admin user or even an admin user at all.

To detect whether or not there is only one admin user left, additional api calls will be needed. You'll need to make a request to the `/api/v2/users` endpoint with query params `{ is_superuser: True }`:

```
http GET /api/v2/users/?is_superuser=True`
```

If the count from that query is `1` _and_ the user you're attempting to modify is a superuser, you could show a warning.

The other thing to remember is that the _current user_ that is performing the update must be a superuser as well, otherwise they have no way of knowing how many total super users exist on the system in the first place. You will need to check for this condition before doing any of these other checks. The overall process is as follows:

```
if the user is modifying themselves:
  if the user is being changed from super user to non super-user:
    make a request to http GET /api/v2/users/?is_superuser=True`
    if the count from this request is 1:
      show a warning
else:
  do exactly the same thing as before
```

_Alternatively_ we could just show the warning whenever a superuser tries to remove their own superuser privilege, regardless of whether or not they're the _last remaining_ superuser. I believe that would ultimately solve the same problem, and would be less complicated to implement:


```
if the user is modifying themselves:
  if the user is being changed from super user to non super-user:
      show a warning
else:
  do exactly the same thing as before
```

jakemcdermott reviewed (changes requested):
Hello! Thanks for the PR.

Only a superuser can know whether or not someone is the last remaining superuser. The best we can do here is solve the following problem:

 "The last remaining superuser on the system should be warned before removing _their own_ superuser privileges".

_Alternatively_, we can simplify things quite a bit by showing a warning if a superuser is trying to remove their own super_user privileges, regardless of whether or not they're the _last remaining_ superuser.

I think either would be fine depending on how specific you want to be. That said, the second option would be less complicated to implement. In my review, I've given some general pseudo-code and an outline of what would need to be done for either approach.

soniya20 committed:
users-edit.controller.js edited

#5522

soniya20 committed:
Changing users.form.js

Changing the files

soniya20 committed:
Updating users-edit.controller.js

Correcting the indentation of if block

soniya20 commented:
Thank you @jakemcdermott for reviewing the PR.

My idea was not based on the last remaining admin user. I had the observation that the id of the admin user who is created at the time of installation of AWX is always 1. Based on this observation, i had disabled the edit-ability of the user type drop down for user with id equal to 1. Thereby, ensuring that the minimum number of admin users in an AWX deployment can only be one.

However, i can easily rollback the change if you suggest. Please let me know

The alternate approach suggested by you is already covered by the pop up message that i have introduced when a user type is changed from administrator to any other user type.

I have incorporated the review comments provided by you on awx/ui/client/src/users/edit/users-edit.controller

nomikh143 reviewed (approved):
Hello
How will it cahnge by itself? And how mush is it reliable and also want to root superuser privilleges need a bit help thnx....

soniya20 commented:
Hello @nomikh143,
User type won't change by itself.It is that if a user is accidentally changed from administrator to other user type, a pop up warning could mitigate the risk. Here, root superuser have access to all the system permissions. 
Also i have rollback the approach of freezing the default user’s permission which have user id ‘1’.

nomikh143 commented:
> Hello @nomikh143,
> User type won't change by itself.It is that if a user is accidentally changed from administrator to other user type, a pop up warning could mitigate the risk. Here, root superuser have access to all the system permissions. 
> Also i have rollback the approach of freezing the default user’s permission which have user id ‘1’.
It says merge not allowed what is this an error or what?```

```

soniya20 commented:
Hello @nomikh143,
I am not able to check the error detail as it is failing in community automated test cases. If you can share the detailed error so i will be able to investigate the error cause.

nomikh143 commented:
Ok but what kind of a repory u need a bug report?

soniya20 commented:
hii @nomikh143, 
Yes, something like that. Details as to why the automated ansible check has failed

ryanpetrello commented:
At this point, I think it's more reasonable to track and implement this in the upcoming UI rewrite (rather than the _current_ UI implementation).

As such, I'm going to close this PR and ask @mabashian and @jakemcdermott that we track this as part of the new UI implementation that's currently in progress:

https://github.com/ansible/awx/issues/5522
